### ISSUE TITLE
[Feature request] Multiple global steps for better tracking loss with various batch size

### ISSUE THREAD

**Issue by tobyclh** [wandb/client#410]
Hello, thanks for all the work.
Would love to see whether it is possible to have a step count for each plot.
This is useful to 
1. compare against different runs from different machines (with different batch size/runtime
2. epoch-based metrics (ROC etc.
3. tracking validation loss during training

**Comment by vanpelt** [comment]
Hey @tobyclh, we increment a counter every time you call `wandb.log`.  If you pass the **step** keyword to `wandb.log` we'll set the step counter to that.  This is the default X axis in all charts on your run page.  If you would like to log additional steps you can just add them to your call to `wandb.log` i.e. `wandb.log({"accuracy": 123, "custom_step": 3})`.  As long as the metric you're logging is monotonically increasing you can change the X-Axis to use that value.  Just click the "pin" icon on a chart, then the "edit" icon (little pencil) to configure the chart.

![Screen Shot 2019-07-02 at 9 49 44 AM](https://user-images.githubusercontent.com/17/60530998-e2832980-9cae-11e9-92b2-7b709810d2d4.png)

**Comment by tobyclh** [comment]
This is godsend. Thanks for telling me that!
Thank you very much!

**Comment by tshrjn** [comment]
Each metric has a separate counter in tensorboard. A possible use case example:
```
self.tb_writer.add_scalar('Loss/Train', running_loss/args.log_interval, epoch * len(self.trainloader) + i)


....
self.tb_writer.add_scalar('Loss/Valid', valid_loss, epoch)
```

However, when using sync_tensorboard, there is only 1 single counter wandb's dashboard & the plots become super difficult to understand. Am I missing something?

**Comment by vanpelt** [comment]
@tshrjn for tensorboard each summary writer has it's own global step.  In the UI you can choose a different x-axis for charts either globally or at the chart level.  If you link me to one of your runs I can look into this specific case.  

![image](https://user-images.githubusercontent.com/17/79612415-80ff4080-80b1-11ea-9bc8-83afc5410e7c.png)

**Comment by tshrjn** [comment]
Hi, thanks for the quick response. I did choose step but it chooses some other global step.
I created a demo showing this [here](https://app.wandb.ai/tshrjn/demo-steps/workspace?workspace=user-tshrjn).


The code for it is as follows:

import wandb
from torch.utils.tensorboard import SummaryWriter

```
wandb.init(project="demo-steps", sync_tensorboard=True)
writer = SummaryWriter()

bs, epochs = 10, 100

for epoch in range(epochs):
    for batch_idx in range(bs):
        writer.add_scalar('batch', batch_idx, epoch * bs + batch_idx)
    writer.add_scalar('epoch', epoch, epoch)
```

The epochs step should only be from 1 to 100. Here y=x & I can choose `epoch` to be the  x-axis. However, usually, that's not the case.

Please let me know if I'm wrong somewhere or need more clarification.

**Comment by tshrjn** [comment]
Is there an update on this?
Looking at the code above & then graphs, it'd be immediately clear that there's some issue with the x-axis steps for epochs.

**Comment by vanpelt** [comment]
Hey @tshrjn the issue you're seeing is caused by having two different global steps.  If you change your example to always use your batch idx as the tensorboard global_step and log the epoch along with it you'll be able to choose between "global_step" and "epoch" as your x-axis in the wandb ui.  You can't choose global_step currently because it's not monotonically increasing.

```python
import wandb
from torch.utils.tensorboard import SummaryWriter

wandb.init(project="demo-steps", sync_tensorboard=True)
writer = SummaryWriter()

bs, epochs = 10, 100

for epoch in range(epochs):
    for batch_idx in range(bs):
        writer.add_scalar('batch', batch_idx, epoch * bs + batch_idx)
    writer.add_scalar('epoch', epoch, epoch * bs + bs - 1)
```

**Comment by tshrjn** [comment]
Hi @vanpelt, thanks for the suggestion. Though I feel it's more of a workaround and not the desired behaviour. This is a simplified example & there might be a lot of situations where we actually want local counters instead of 1 global one.

If you check the tensorboard for the above example, you will see it maintains separate counter/step for each metric (here scalars) & that's the behaviour which should be shown here as well.
I.e. each variable/metric having its independent counter especially when importing/syncing with tensorboard (but I think also in general). 

Even in general, perhaps an option for local step should be provided. What do you think?

**Comment by vanpelt** [comment]
@tshrjn Let me clarify.  The tensorboard integration takes calls like`add_scalar(key, value, global_step)` and converts them into `wandb.log({"key": value, "global_step": global_step})` calls.  This enables us to "rollup" multiple calls to `add_scalar` with the same global_step so the final `wandb.log` call would contain all keys for the same step, i.e. `wandb.log({"key1": value1, "key2": value2, "global_step": global_step})`.  In addition if you have multiple SummaryWriters that write to different tfevent files we namespace the global steps so that you can have 2 different global_steps.

The reason we rollup metrics across "global_step" is to allow you to plot metrics that share the same step on the same plot.  In tensorboard for instance you could never plot "batch" and "epoch" on the same plot using the global_step axis.

We're working on an overhaul of our metrics backend to enable updating steps in the past.  Until then it's best to make your "global_steps" only increase during training.

**Comment by vanpelt** [comment]
Also, circling back to you mentioning an option for a local step.  If any of the keys that get rolled up into the `wandb.log` calls are monotonically increasing during training you can use them as x-axis in wandb essentially making them local steps.  So if you use a universal global_step  that is always increasing you could run the following:

```python
bs, epochs, global_step = 10, 100, 0
for epoch in range(epochs):
    for batch_idx in range(bs):
        global_step += 1
        writer.add_scalar('batch_acc', acc, global_step)
    writer.add_scalar('epoch_step', epoch // bs, global_step)
    writer.add_scalar('epoch_acc', acc, global_step)
```

This would enable you to plot "batch_acc" and "epoch_acc" on the "epoch_step" x-axis along with the "global_step" x-axis.
